ttnn/cpp/ttnn/operations/normalization/softmax/softmax_compiler_interface.hpp

```diff
@@ -0,0 +1,59 @@
+// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include <exception>
+#include <functional>
+
+#include "third_party/json/json.hpp"
+#include "tt_metal/common/logger.hpp"
+#include "ttnn/compiler_interface/compiler_interface.hpp"
+#include "ttnn/graph/graph_processor.hpp"
+#include "ttnn/graph/graph_trace_utils.hpp"
+#include "ttnn/operations/normalization/softmax/softmax.hpp"
+
+namespace ttnn::compiler_interface {
+
+QueryResponse softmax_op_constraints(
+    Device* device, const TensorSpec& input, const int dim_arg, const TensorSpec& output) {
+    // get_op_trace is a lambda that prepares input and output tensors, capturing graph trace of the op without
+    // inputs allocation.
+    auto get_op_trace = [](Device* device, const TensorSpec& input, const int dim_arg, const TensorSpec& output) {
+        nlohmann::json op_trace;
+
+        // outer graph capture is used to avoid capturing and dispatching of dummy input tensor(s) creation
+        {
+            ttnn::graph::GraphProcessor::begin_graph_capture(ttnn::graph::GraphProcessor::RunMode::NO_DISPATCH);
+            const auto input_tensor = create_device_tensor(input, device);
+
+            // output tensor is created in the inner graph capture to capture its allocation
+            {
+                ttnn::graph::GraphProcessor::begin_graph_capture(ttnn::graph::GraphProcessor::RunMode::NO_DISPATCH);
+                auto output_tensor = ttnn::softmax(input_tensor, dim_arg, output.tensor_layout().get_memory_config());
+                // close inner graph capture, before output buffer is deallocated
+                op_trace = ttnn::graph::GraphProcessor::end_graph_capture();
+            }
+            // close outer graph capture
+            ttnn::graph::GraphProcessor::end_graph_capture();
+        }
+
+        // TODO(mbezulj) remove this debug print
+        // std::cout << graph::to_graphviz(op_trace) << std::endl;
+
+        return op_trace;
+    };
```

Comment:
Too many repetitions of the same code.

Please refactor to do the graph capture in a single location.

Maybe invert dependencies (one method, pass in method to do what's inside for each op) or maybe have these as part of a class/interface/struct that calls common code and then branches.

Also, may have these methods be op_constraints, and then pass in a lambda/struct that has the op specific code/parameters.

Reply:
thanks for the feedback, let me try with inversion

Reply:
@bbradelTT  please take a look at the change. I removes code repetitions. I've kept per op interface declarations so we have a clean line of what we test and what compiler is going to use.